### PR TITLE
Support globstar (**) syntax in paths include/exclude with wcmatch.

### PR DESCRIPTION
**Pull request by Isaac0616**
In #2851, I tried to use `pathspec` to support globstar (**) syntax (#1834).
However, there is a bug which prevents matching absolute paths and I can't get the library to fix it.
I tried another library `wcmatch` to support the globstar syntax in this pull request.
[`wcmatch`](https://facelessuser.github.io/wcmatch/) is more actively maintained and supports similar APIs/semantic to standard library's `glob` and `pathlib`.
It makes this updates backward compatible (supposed to be, mostly).
(In contrast to `pathspec`, which is slightly different and follows gitignore's semantics.)

In this pull request, I only changed the `match_glob` function.
This is probably the easiest way / minimum changes to support the case that pattern `foo/bar` should match paths `x/foo/bar`, `foo/bar/x`, and `x/foo/bar/x`.
(`wcmatch` follows bash's semantics, so the pattern should be `**/foo/bar/**` to natively match `x/foo/bar`, `foo/bar/x`, and `x/foo/bar/x`).
It is not very clean and might have performance impact (hopefully not too bad compare to the original).
I can see some complexity comes from that we have to shell out to git to respect gitignore.
I am thinking whether there is a better way to implement `TargetManager`, but if this pull request is acceptable, that can be a future work.
I will also have to figure out a good benchmark for such changes.

### Test plan
I updated the unit test to add the test cases for globstar and question mark and make sure the previous tests still pass.
I also made the unit tests to compare the actual content of the `Set` instead of only comparing the length of the `Set`.
(Some e2e tests failed, but they seems to fail before this pull request.)

### CODE REVIEW

**Comment by Isaac0616** [comment]
I tried to be backward compatible and made pattern `foo/bar` matches paths `x/foo/bar`, `foo/bar/x`, and `x/foo/bar/x`.
In other words, there are implicit `**` before and after the pattern (more precisely, `foo/bar` is `**/foo/bar` + `**/foo/bar/**`).
We can discuss whether we should keep the behavior.
Requiring explicit `**` before the pattern allow `foo/bar` to match only from the root folder, which seems useful.
For the `**` after the pattern, I can see that it is handy if the pattern is a directory.

File: semgrep/semgrep/target_manager.py [hunk 1]
```diff
@@ -211,7 +217,12 @@ def filter_includes(arr: Set[Path], includes: List[str]) -> Set[Path]:
     def filter_excludes(arr: Set[Path], excludes: List[str]) -> Set[Path]:
         """
         Returns all elements in arr that do not match any excludes pattern
+
+        If includes is empty, returns arr unchanged
```
Comment (brendongo) [inline semgrep/semgrep/target_manager.py]:
```suggestion
        If excludes is empty, returns arr unchanged
```

File: semgrep/setup.py [hunk 2]
```diff
@@ -108,6 +108,7 @@ def find_executable(env_name, exec_name):
         "tqdm>=4.46.1",
         "packaging>=20.4",
         "jsonschema~=3.2.0",
+        "wcmatch>=8.2",
```
Comment (brendongo) [inline semgrep/setup.py]:
```suggestion
        "wcmatch==8.2",
```
Can we pin to a specific version of wcmatch for now. A change in file targeting behavior would have an outsized impact on semgrep behavior.

**Review by brendongo — COMMENTED** [review]
Thanks for the PR @Isaac0616 looks good. Really appreciate the tests. Something idk if you've considered is what happens to glob matching when a user runs `semgrep --config xyz --exclude */bar/* foo` and it turns out the full path of foo is `/a/b/c/bar/d/e/foo`?

**Comment by Isaac0616** [comment]
Hi @brendongo,
Sorry for the delay, I was working on something with a tight deadline.
I updated the commit according to your comments.
Regarding the question of `semgrep --config xyz --exclude '**/bar/**' foo`, `foo` won't be excluded because we are providing the relative path.
If we on the other hand use `semgrep --config xyz --exclude '**/bar/**' /a/b/c/bar/d/e/foo`, `foo` will be excluded.
I think this behavior is consistent without the changes in this commit.
That is, `semgrep --config xyz --exclude 'bar' /a/b/c/bar/d/e/foo` will exclude `foo` but `semgrep --config xyz --exclude 'bar' foo` won't (`bar` implies `**/bar/**` in our current setting).
Does it answer your question?

File: semgrep/semgrep/target_manager.py [hunk 3]
```diff
@@ -193,7 +194,12 @@ def match_glob(path: Path, globs: List[str]) -> bool:
         Return true if path or any parent of path matches any glob in globs
         """
         subpaths = [path, *path.parents]
-        return any(p.match(glob) for p in subpaths for glob in globs)
+        return any(
+            wcpathlib.PurePath(p).match(
+                globs, flags=wcpathlib.GLOBSTAR | wcpathlib.DOTGLOB
+            )
+            for p in subpaths
+        )
```
Comment (Isaac0616) [inline semgrep/semgrep/target_manager.py]: wcmatch's `PurePath.match` is slower and looping like this is quite a performance hit.
Let me try another approach.

File: semgrep/semgrep/target_manager.py [hunk 4]
```diff
@@ -188,12 +189,21 @@ def expand_targets(
         return expanded
 
     @staticmethod
-    def match_glob(path: Path, globs: List[str]) -> bool:
+    def preprocess_path_patterns(patterns: List[str]) -> List[str]:
```
Comment (Isaac0616) [inline semgrep/semgrep/target_manager.py:193]: @brendongo 
Because calling `wcmatch`'s `PurePath.match` for each subpaths is too slow, I switched to this approach.
Let me know what do you think.
It looks a bit hacky, but I believe we can use unit tests to ensure it works as expected.

Comment (brendongo) [reply semgrep/semgrep/target_manager.py:193]: Yeah looks good to me!

**Comment by brendongo** [comment]
@Isaac0616 thanks for the PR it looks great! Do you mind fixing the conflict and we can get this merged

**Review by brendongo — APPROVED** [review]

**Comment by mjambon** [comment]
Hi, we're planning to revert some of this behavior to get closer to what git does. See https://github.com/returntocorp/semgrep/issues/6835

tl;dr according to the [gitignore spec](https://git-scm.com/docs/gitignore) `a/b` is anchored and is equivalent to `/a/b` while `a` is unanchored and equivalent to `**/a`. we will implement this.
